🎁 Add `search` to file location methods in tasks

babakks opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

Fixes #160771

A new `fileLocation` method, named `search`, is added to problem matches. The usage of this file location method is as:

```jsonc
"problemMatcher": {
    "fileLocation": [
        "search",
        {
            "include": [ // (Required)
                "${workspaceFolder}/subdir1",
                "${workspaceFolder}/subdir2"
            ],
            "exclude": [ // (Optional)
                "${workspaceFolder}/subdir1/subdir-to-ignore",
                "${workspaceFolder}/subdir2/subdir-to-ignore"
            ]
        }
    ]
```

The `include` field is required, but the `exclude` is not.

To verify this PR, you can use [babakks/vscode-verify-search-file-location](https://github.com/babakks/vscode-verify-search-file-location) repo as the test workspace.

meganrogge reviewed (approved):
Very nice 👏🏼

alexr00 reviewed (commented):
I didn't see variables in the `"includes"` and `"excludes"` section get resolved when testing this.

babakks commented on src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:
```diff
@@ -1719,7 +1725,15 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 				if (hasFilePrefix) {
-					copy.filePrefix = await this._resolveVariable(resolver, copy.filePrefix);
+					if (Types.isString(matcher.filePrefix)) {
+						copy.filePrefix = await this._resolveVariable(resolver, copy.filePrefix as string);
```
Applied.

babakks commented on src/vs/workbench/contrib/tasks/common/problemMatcher.ts:
```diff
@@ -212,6 +215,18 @@ export async function getResource(filename: string, matcher: ProblemMatcher, fil
 		matcherClone.fileLocation = FileLocationKind.Absolute;
 		return getResource(filename, matcherClone);
+	} else if (kind === FileLocationKind.Search && fileService) {
+		const fsProvider = fileService.getProvider('file');
```
Applied.

babakks commented on src/vs/workbench/contrib/tasks/common/problemMatcher.ts:
```diff
@@ -1612,9 +1678,24 @@ export namespace Schemas {
 					}
 				],
-				description: localize('ProblemMatcherSchema.fileLocation', 'Defines how file names reported in a problem pattern should be interpreted. A relative fileLocation may be an array, where the second element of the array is the path the relative file location.')
+				description: localize('ProblemMatcherSchema.fileLocation', 'Defines how file names reported in a problem pattern should be interpreted. A relative fileLocation may be an array, where the second element of the array is the path of the relative file location.')
```
Added to the description.

babakks commented on src/vs/workbench/contrib/tasks/common/problemMatcher.ts:
```diff
@@ -1612,9 +1678,24 @@ export namespace Schemas {
+							{
+								type: 'object',
+								properties: {
+									'include': { type: 'array', items: { type: 'string' } },
```
Added examples to the schema which will be rendered as suggestions in the runtime.

Also, for a better experience, changed the type of `include`/`exclude` arrays to accept single values as well; i.e., they're now `string | string[]`.

babakks replied:
But, I'm now sure why validation didn't work.

babakks commented:
@alexr00 I've also added two other examples for the `relative` and `autoDetect` file location methods.

babakks commented:
> I didn't see variables in the `"includes"` and `"excludes"` section get resolved when testing this.

@alexr00 I'm not sure I'm getting your point, correctly. When I tested, the variables (e.g., `"${workspaceFolder}"`) were replaced with their values via the `TerminalTaskSystem._resolveVariable` method.

babakks commented:
@meganrogge @alexr00 Couldn't ask you both for the review from the side links, so could you please re-review this?

babakks commented:
@meganrogge @alexr00 We can also make the `include` property optional (default to `${workspaceFolder}`). What do you think?

babakks commented:
It had a shortcoming that when the `filename` was more than a mere filename and extension (e.g., 'subdir/name.ext`). Fixed that in the last commit; also updated the verification repo to include new tasks for this purpose.

meganrogge commented:
> @meganrogge @alexr00 We can also make the `include` property optional too (default to `${workspaceFolder}`). What do you think?

I like that idea - thanks for working on this

babakks commented:
The `include` property is now optional and its default is `"${workspaceFolder}"`.
